xla/service/gpu/transforms/cudnn_sdpa_score_mod.cc

```diff
@@ -0,0 +1,366 @@
+    case dnn::DataType::kInt8:
+      return cudnn_frontend::DataType_t::INT8;
+    case dnn::DataType::kInt32:
+      return cudnn_frontend::DataType_t::INT32;
+    case dnn::DataType::kInt64:
+      return cudnn_frontend::DataType_t::INT64;
+    case dnn::DataType::kBF16:
+      return cudnn_frontend::DataType_t::BFLOAT16;
+    case dnn::DataType::kF8E4M3FN:
+      return cudnn_frontend::DataType_t::FP8_E4M3;
+    case dnn::DataType::kF8E5M2:
+      return cudnn_frontend::DataType_t::FP8_E5M2;
+#if CUDNN_VERSION >= 90700
+    case dnn::DataType::kF4E2M1FN:
+      return cudnn_frontend::DataType_t::FP4_E2M1;
+    case dnn::DataType::kF8E8M0FNU:
+      return cudnn_frontend::DataType_t::FP8_E8M0;
+#endif
+    default:
+      LOG(FATAL) << "Invalid DNN data type: " << static_cast<int>(data_type);
+  }
+}
+
+template <xla::PrimitiveType XlaT, typename T>
+Tensor ScoreModFunc::LiteralToCudnnTensor(const xla::HloInstruction* hlo,
+                                          cudnn_frontend::graph::Graph& graph) {
+  using NativeT =
+      typename xla::primitive_util::PrimitiveTypeToNative<XlaT>::type;
+  return graph.tensor(T(hlo->literal().GetFirstElement<NativeT>()));
+}
+
+absl::Status ScoreModFunc::UpdateCudnnMap(cudnn_frontend::graph::Graph& graph,
+                                          Uid next_uid) {
+  TF_RETURN_IF_ERROR(UpdateHloParameterToCudnnMap(graph, fwd_hlo_to_cudnn_,
+                                                  fwd_comp_, next_uid));
+  TF_RETURN_IF_ERROR(
+      UpdateHloConstantToCudnnMap(graph, fwd_hlo_to_cudnn_, fwd_comp_));
+  if (bwd_comp_) {
+    TF_RETURN_IF_ERROR(
+        UpdateHloConstantToCudnnMap(graph, bwd_hlo_to_cudnn_, bwd_comp_));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status ScoreModFunc::UpdateHloParameterToCudnnMap(
+    cudnn_frontend::graph::Graph& graph,
+    absl::flat_hash_map<const xla::HloInstruction*, Tensor>& hlo_to_cudnn,
+    const xla::HloComputation* computation, Uid next_uid) {
+  for (int i = 1; i < computation->num_parameters(); i++) {
+    auto parameter = computation->parameter_instruction(i);
+    TF_ASSIGN_OR_RETURN(const dnn::DataType type,
+                        xla::gpu::GetDNNDataTypeFromPrimitiveType(
+                            parameter->shape().element_type()));
+    auto desc = dnn::TensorDescriptor::For(
+        type, parameter->shape().dimensions(),
+        parameter->shape().layout().minor_to_major());
+    auto dims = desc.dimensions();
+    auto strides = desc.GetLogicalStrides();
+    auto rank = dims.size();
+    for (int i = 0; i < 4 - rank; i++) {
+      // Pad dims and strides to rank 4
+      dims.push_back(1);
+      strides.push_back(1);
+    }
+    hlo_to_cudnn[parameter] = graph.tensor(
+        cudnn_frontend::graph::Tensor_attributes()
+            .set_dim(dims)
+            .set_stride(strides)
+            .set_data_type(ToCudnnFrontendDataType(desc.type()))
+            .set_name(absl::StrCat("score_mod_input_", std::to_string(i)))
```

Comment:
Nit: No need to `std::to_string` since `absl::StrCat` can take integers as well.